pkg/controller/snapshotter/snapshotter.go

```diff
@@ -347,204 +345,310 @@ func (vs *volumeSnapshotter) deleteSnapshot(spec *v1.PersistentVolumeSpec, sourc
+				tags[CloudSnapshotCreatedForVolumeSnapshotNamespaceTag] = snapshot.Metadata.Namespace
+				tags[CloudSnapshotCreatedForVolumeSnapshotNameTag] = snapshot.Metadata.Name
+				tags[CloudSnapshotCreatedForVolumeSnapshotTimestampTag] = timestamp
+				glog.Infof("findVolumeSnapshotMetadata: returning tags [%#v]", tags)
 			}
+		}
+	}
+	return tags
+}
 
-			// If metadata not found, create it in VolumeSnapshot now
-			if !bFoundVolumeSnapshotMetadata {
-				tags, err = vs.UpdateVolumeSnapshot(snapshotName, nil, bFoundVolumeSnapshotMetadata)
-			}
+func (vs *volumeSnapshotter) getPlugin(snapshotName string, snapshot *crdv1.VolumeSnapshot) (*volume.VolumePlugin, error) {
+	pv, err := vs.getPVFromVolumeSnapshot(snapshotName, snapshot)
+	if err != nil {
+		return nil, err
+	}
+	spec := &pv.Spec
+	volumeType := crdv1.GetSupportedVolumeFromPVSpec(spec)
+	if len(volumeType) == 0 {
+		return nil, fmt.Errorf("Unsupported volume type found in PV %#v", spec)
+	}
+	plugin, ok := (*vs.volumePlugins)[volumeType]
+	if !ok {
+		return nil, fmt.Errorf("%s is not supported volume for %#v", volumeType, spec)
+	}
+	return &plugin, nil
+}
 
-			var snapshotDataSource *crdv1.VolumeSnapshotDataSource = nil
-			var snapStatus *[]crdv1.VolumeSnapshotCondition = nil
-			// Look for snapshot by the plugin if metadata already existed in VolumeSnapshot
-			if bFoundVolumeSnapshotMetadata {
-				spec := &pv.Spec
-				volumeType := crdv1.GetSupportedVolumeFromPVSpec(spec)
-				if len(volumeType) == 0 {
-					return fmt.Errorf("Unsupported volume type found in PV %#v", spec)
-				}
-				plugin, ok := (*vs.volumePlugins)[volumeType]
-				if !ok {
-					return fmt.Errorf("%s is not supported volume for %#v", volumeType, spec)
-				}
+/****************************************************************************************
+*func (vs *volumeSnapshotter) getSnapshotStatus(snapshot *crdv1.VolumeSnapshot) (string, *crdv1.VolumeSnapshotData, error) {
+*	var snapshotName string = snapshot.Metadata.Name
+*	var snapshotDataObj *crdv1.VolumeSnapshotData = nil
+*	status = getSimplifiedSnapshotStatus(snapshot.Status.Conditions)
+*	if status == statusFinished || status == statusPending {
+*		// If we are here, takeSnapshot has already happened.
+*		// Check whether the VolumeSnapshotData object is already created
+*		snapshotDataObj := vs.getSnapshotDataFromSnapshotName(snapshotName)
+*		if snapshotDataObj == nil {
+*			// TODO: Create VolumeSnapshotData obj
+*		}
+*		return status, snapshotDataObj, nil
+*	} else {
+*		tags :=  vs.findVolumeSnapshotMetadata(snapshot)
+*		if tags != nil {
+*			plugin, err := vs.getPlugin(snapshotName, snapshot)
+*			if plugin == nil {
+*				glog.Errorf("Failed to get volume plugin.")
+*				return status, nil, err
+*			}
+*			// Check whether the real snapshot is already created by the plugin
+*			glog.Infof("getSnapshotCreateFunc: FindSnapshot %s ...", snapshotName)
+*			snapshotDataSource, snapStatus, err = plugin.FindSnapshot(tags)
+*			if err != nil {
```

Comment:
The logic from here is that if snapshot is found (in this case, error is nil, and it should also return snapshotDataSource and snapStatus at the same time), we can go ahead update the VolumeSnapshot status, and also create VolumeSnapshotData object. 
Otherwise, just return status (which is statusNew)

Reply:
if err == nil {
    // updateVolumeStatusShot with status
   // create volumeSnapshotData object and bind them
}

you don't need to check whether snapStatu is nil or not (it must be not nil). Also VolumeSnapshotData object is also not created yet at this point. Otherwise the snapshot status will be pending. You can directly create VolumeSnapshotData and bind them